connect/api/src/main/java/org/apache/kafka/connect/source/SourceRecord.java

```diff
@@ -46,25 +46,49 @@
     private final Map<String, ?> sourcePartition;
     private final Map<String, ?> sourceOffset;
 
+    /**
+     * @deprecated use {@link SourceRecord.Builder}
+     */
+    @Deprecated
```

Comment:
I think I mentioned this offline, but I'm not sure I would deprecate this yet. I think there are a number of tradeoffs here, this is the set that come to mind immediately:

Pros:
* One builder class, although it has a bunch of per-field methods, will beat the combinatoric explosion of possible constructors when the number of fields increases (as it is now).
* Less error prone for users as they don't have to try to dig out the right constructor to use that has just the right set of fields. If there are any fields with the same type (schema and key/value fields are the ones that come to mind), users may accidentally pass the wrong variable as that argument and not be caught by the compiler.
* Easier compatibility story -- as long as the setter for each field continues to exist in the `Builder`, we're good.

Cons:
* Deprecating relatively new APIs. May be more painful for us to maintain them, but we probably shouldn't be overly aggressive since we want folks to have a stable API to work with.
* Overhead of builder -- I think in a lot of cases this is not a concern as the object is so short lived that collecting it will be very cheap. However, consider that this *can* be very hot code -- in some cases creating the SourceRecord might be *the majority* of the work involved in moving messages (e.g. Schema is fixed and a primitive type, so there's not even any other conversion happening). Is the expectation that those use cases would use the single all-fields SourceRecord constructor directly?
* How many variants of the constructor are realistic? We're definitely not expecting to add more fields to messages (that type of change has to come with a *huge* functionality win to ever make it past KIPs). Also, at least a few fields only make sense when combined (e.g. `valueSchema` + `value`). I'm not sure the number of variants is that high (but maybe I'm wrong -- I'm guessing you raised this as you started copy/pasting all the existing ones to include timestamp).

Heh, not great when my pros/cons list ends up with an equal number of items on either side, but these issues probably warrant further discussion. (I do also very much appreciate keeping an eye out for ways we can cleanup/simplify the API and make the connector developer experience better.)